docs/design/in-doc-search-ui.rst

```diff
@@ -0,0 +1,183 @@
+This design document aims to provides the details of it.
+This is a GSoC'19 project.
+
+.. warning::
+
+    This design document details future features that are **not yet implemented**.
+    To discuss this document, please get in touch in the `issue tracker`_.
+
+
+The final result may look something like this:
+
+.. figure:: ../_static/images/design-docs/in-doc-search-ui/in-doc-search-ui-demo.gif
+    :align: center
+    :target: ../_static/images/design-docs/in-doc-search-ui/in-doc-search-ui-demo.gif
+
+    Short demo
+
+
+Goals And Non-Goals
+-------------------
+
+Project Goals
+++++++++++++++
+
+* Support a search-as-you-type/autocomplete interface.
+* Support across all (or virtually all) Sphinx themes.
+* Support for the JavaScript user experience down to IE11 or graceful degradation where we can't support it.
+* Project maintainers should have a way to opt-in/opt-out of this feature.
+* (Optional) Project maintainers should have the flexibility to change some of the styles using custom CSS and JS files.
+
+Non-Goals
+++++++++++
+
+* For the initial release, we are targeting only Sphinx documentations
+  as we don't index MkDocs documentations to our Elasticsearch index.
+
+
+Existing Search Implementation
+------------------------------
+
+We have a detailed documentation explaing the underlying architecture of our search backend
+and how we index documents to our Elasticsearch index.
+You can read about it :doc:`here <../development/search>`.
+
+
+Proposed Architecture for In-Doc Search UI
+------------------------------------------
+
+Frontend
+++++++++
+
+Technologies
+~~~~~~~~~~~~
+
+Frontend is to designed in a theme agnostics way. For that,
+we explored various libraries which may be of use but none of them fits our needs.
+So, we might be using vanilla JavaScript for this purpose.
+This will provide us some advantages over using any third party library:
+
+* Better control over the DOM.
+* Performance benefits.
+
+
+Proposed Architecture
+~~~~~~~~~~~~~~~~~~~~~
+
+We plan to select the search bar, which is present in every theme,
+and use the `querySelector()`_ method of JavaScript.
+Then add an event listener to it to listen for the changes and
+fire a search query to our backend as soon as there is any change.
+Our backend will then return the suggestions,
+which will be shown to the user in a clean and minimal UI.
+We will be using `document.createElement()`_ and `node.removeChild()`_ method
+provided by JavaScript as we don't want empty `<div>` hanging out in the DOM.
+
+We have a few ways to include the required JavaScript and CSS files in all the projects:
+
+* Add CSS into `readthedocs-doc-embed.css` and JS into `readthedocs-doc-embed.js`
+  and it will get included.
+* Package the in-doc search into it's own self-contained CSS and JS files
+  and include them in a similar manner to `readthedocs-doc-embed.*`.
+* It might be possible to package up the in-doc CSS/JS as a sphinx extension.
```

Comment:
If this would be a voting, I'd vote for this one :)

Reply:
I would also vote for that one. We are currently in this direction only.

Reply:
My only concern with the extension is that it _doesn't work locally_. You can build your docs locally, but even if we hit the prod API, it will only return results from prod docs. This feels like a weird thing to do with an extension when it's inherently linked to our RTD production builds.

Reply:
@ericholscher 
We have a global object `READTHEDOCS_DATA` and it contains `api_host`. In the extension, we are taking the API host from there. 
So in local, it is - http://127.0.0.1:8000
And we can use the extension locally (given that the Elasticsearch is correctly set up).

Reply:
Sure, but that requires a running instance of RTD -- I mean it isn't useful outside of RTD. If you already have a local RTD instance, then bundling it with RTD would have the same outcome :)

Reply:
Yes... that is the limitation. :confused:

Reply:
Split of code is a win to me. Easier to contribute, easier to focus when debugging, easier to keep up to date, etc. The downside is deploying a bugfix immediately, where core & extension need to be deployed together.

I understand that could feel weird that the results come from production while you are seeing a different set of docs. The extension could disable itself if detecting that it's running outside RTD if we can avoid that weird UX. Not a solution, though.

Reply:
I think it will increase complexity in development environment..!!